eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// <auto-generated/>
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace BasicTypeSpec
+{
+    internal partial class BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT : AsyncPageable<ThingModel>
+    {
+        private readonly BasicTypeSpecClient _client;
+        private readonly string _token;
+        private readonly RequestContext _context;
+
+        /// <summary> Initializes a new instance of BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT, which is used to iterate over the pages of a collection. </summary>
+        /// <param name="client"> The BasicTypeSpecClient client used to send requests. </param>
+        /// <param name="token"></param>
+        /// <param name="context"> The request options, which can override default behaviors of the client pipeline on a per-call basis. </param>
+        public BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT(BasicTypeSpecClient client, string token, RequestContext context)
+        {
+            _client = client;
+            _token = token;
+            _context = context;
+        }
+
+        /// <summary> Gets the pages of BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT as an enumerable collection. </summary>
+        /// <param name="continuationToken"> The continuation token. </param>
+        /// <param name="pageSizeHint"> The page size hint. </param>
+        /// <returns> The pages of BasicTypeSpecClientListWithContinuationTokenAsyncCollectionResultOfT as an enumerable collection. </returns>
+        public override async IAsyncEnumerable<Page<ThingModel>> AsPages(string continuationToken, int? pageSizeHint)
+        {
+            string nextLink = continuationToken;
+            do
+            {
+                Response response = await GetNextResponse(pageSizeHint, nextLink).ConfigureAwait(false);
+                if (response is null)
+                {
+                    yield break;
+                }
+                ListWithContinuationTokenResponse items = (ListWithContinuationTokenResponse)response;
+                nextLink = items.NextToken.ToString();
+                yield return Page<ThingModel>.FromValues((IReadOnlyList<ThingModel>)items.Things, nextLink, response);
```

Comment:
https://github.com/microsoft/typespec/issues/7387 is causing extra casting to IReadOnlyList<>